enterprise/ha/src/test/java/org/neo4j/kernel/ha/UpdatePullerTriggersPageTransactionTrackingIT.java

```diff
@@ -0,0 +1,224 @@
+import org.neo4j.io.pagecache.tracing.cursor.context.VersionContext;
+import org.neo4j.io.pagecache.tracing.cursor.context.VersionContextSupplier;
+import org.neo4j.kernel.configuration.Config;
+import org.neo4j.kernel.ha.factory.HighlyAvailableEditionModule;
+import org.neo4j.kernel.impl.context.TransactionVersionContextSupplier;
+import org.neo4j.kernel.impl.factory.DatabaseInfo;
+import org.neo4j.kernel.impl.factory.GraphDatabaseFacade;
+import org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory;
+import org.neo4j.kernel.impl.factory.PlatformModule;
+import org.neo4j.kernel.impl.ha.ClusterManager;
+import org.neo4j.kernel.impl.transaction.log.TransactionIdStore;
+import org.neo4j.test.ha.ClusterRule;
+
+import static org.junit.Assert.assertEquals;
+import static org.neo4j.helpers.collection.MapUtil.stringMap;
+import static org.neo4j.kernel.configuration.Settings.TRUE;
+
+public class UpdatePullerTriggersPageTransactionTrackingIT
+{
+    @Rule
+    public final ClusterRule clusterRule = new ClusterRule( getClass() );
+    private final Label NODE_LABEL = Label.label( "mark" );
+    private final TestTransactionVersionContextSupplier contextSupplier = new TestTransactionVersionContextSupplier();
+    private ClusterManager.ManagedCluster cluster;
+
+    @Before
+    public void setup() throws Exception
+    {
+        CustomGraphDatabaseFactory customGraphDatabaseFactory = new CustomGraphDatabaseFactory();
+        cluster = clusterRule.withSharedSetting( GraphDatabaseSettings.snapshot_query, TRUE )
+                .withDbFactory( customGraphDatabaseFactory )
+                .startCluster();
+        HighlyAvailableGraphDatabase master = cluster.getMaster();
+        for ( int i = 0; i < 3; i++ )
+        {
+            try ( Transaction tx = master.beginTx() )
+            {
+                master.createNode( NODE_LABEL );
+                tx.success();
+            }
+        }
+        cluster.sync();
+    }
+
+    @Test
+    public void updatePullerTriggerPageTransactionTracking()
+    {
+        HighlyAvailableGraphDatabase slave = cluster.getAnySlave();
+        TransactionIdStore slaveTransactionIdStore =
+                slave.getDependencyResolver().resolveDependency( TransactionIdStore.class );
+        assertEquals( 5, slaveTransactionIdStore.getLastClosedTransactionId() );
+
+        ByzantineLongSupplier byzantineIdSupplier = contextSupplier.getByzantineIdSupplier();
+        byzantineIdSupplier.useWrongTxId();
+        try ( Transaction ignored = slave.beginTx() )
+        {
+            slave.execute( "match (n) return n" );
+        }
+        catch ( QueryExecutionException executionException )
+        {
+            assertEquals( "Unable to get clean data snapshot for query 'match (n) return n' after 5 attempts.", executionException.getMessage());
+        }
+        byzantineIdSupplier.useCorrectTxId();
+        slave.execute( "match (n) return n" ).close();
```

Comment:
here we check that update pulling together with starting with tx on slave works as expected:
we check first that in case if our incorrect tx id provider will supply wrong tx id we will not gonna be able to execute query and in this line we check that as soon as id provider supply correct transaction id query execution works as expected